### PR TITLE
Sketch what the bespoke Wayland extensions API might look like

### PR DESCRIPTION
**Pull request by AlanGriffiths**
I think this is the minimal interface needed to support some bespoke Wayland extension implementation by a shell. I'm thinking of something like xdg_decoration_unstable_v1.

@wmww any thoughts?

### CODE REVIEW

File: examples/mir_demo_server/server_example.cpp [hunk 1]
```diff
@@ -150,7 +150,9 @@ try
         me::add_glog_options_to,
         miral::StartupInternalClient{spinner},
         miral::X11Support{},
-        miral::WaylandExtensions{"wl_shell:xdg_wm_base:zxdg_shell_v6:zwlr_layer_shell_v1"},
+        with_extension(miral::WaylandExtensions{"wl_shell:xdg_wm_base:zxdg_shell_v6:zwlr_layer_shell_v1"},
+            "xdg_decoration_unstable_v1", [](wl_display*) { return std::shared_ptr<void>{}; },
+            "some_other_extension", [](wl_display*) { return std::shared_ptr<void>{}; }),
```
Comment (wmww) [inline examples/mir_demo_server/server_example.cpp]: This seems unnecessarily complicated on both sides of the API. Does it make sense to combine the default extensions and the custom extensions in the same function? Would it make more sense to ask the user for a single `WaylandExtensionBuilder`, which creates all the extensions once a `wl_display` is provided?

File: include/miral/miral/wayland_extensions.h [hunk 2]
```diff
@@ -54,7 +57,25 @@ class WaylandExtensions
...
+/// Add a bespoke Wayland extension
+/// \remark Since MirAL 2.5
+auto with_extension(
+    WaylandExtensions const& wayland_extensions,
+    std::string const& name, std::function<std::shared_ptr<void>(wl_display*)> builder) -> WaylandExtensions;
+
+template<typename... Extensions>
+auto with_extension(WaylandExtensions const& wayland_extensions,
+                   std::string const& name, std::function<std::shared_ptr<void>(wl_display*)> builder,
+                   Extensions... extensions) -> WaylandExtensions
```
Comment (wmww) [inline include/miral/miral/wayland_extensions.h]: Do the variadic template arguments have some advantage over just taking a `std::initializer_list`?

**Comment by RAOF** [comment]
Hm. What do we get from this API? As far as I can tell, Miral users already have access to `mir::Server::run_on_wayland_display(std::function<void(wl_display*)>)`, which is all they need to register an extension.

I guess users get automatic lifetime management of any extension data with the `std::shared_ptr<void>`?

I think the interesting part of bespoke extension support is the necessary mapping between Wayland protocol objects and Mir(al) objects, which I'm not sure this API gets us closer to?

**Comment by AlanGriffiths** [comment]
> Hm. What do we get from this API? As far as I can tell, Miral users already have access to `mir::Server::run_on_wayland_display(std::function<void(wl_display*)>)`, which is all they need to register an extension.

One of us is missing something - that wl_display* belongs to the client API. Surely clients should not be able to register extensions?

> I think the interesting part of bespoke extension support is the necessary mapping between Wayland protocol objects and Mir(al) objects, which I'm not sure this API gets us closer to?

Agreed, it doesn't provide a mapping, but this is a first iteration that allows some meaningful protocols to be implemented.

**Comment by RAOF** [comment]
On 6 February 2019 6:35:27 pm AEDT, Alan Griffiths <notifications@github.com> wrote:
>> Hm. What do we get from this API? As far as I can tell, Miral users
>already have access to
>`mir::Server::run_on_wayland_display(std::function<void(wl_display*)>)`,
>which is all they need to register an extension.
>
>One of us is missing something - that wl_display* belongs to the client
>API. Surely clients should not be able to register extensions?

No? That calls WaylandConnector::run_on_wayland_display which runs the functor with the server-side wl_display. That's what shells need to register extensions against.

Convenient how the client and server types are indistinguishable (except by runtime crashes).

**Comment by AlanGriffiths** [comment]
> No? That calls WaylandConnector::run_on_wayland_display which runs the functor with the server-side wl_display. That's what shells need to register extensions against.

Oops! That's what comes of responding too quickly.

`mir::Server::run_on_wayland_display(std::function<void(wl_display*)>)` is NOT exposed through the miral API, just the libmirserver API. (It would be one way to implement the API proposed here.)

I was conflating it with the MirAL API that does exist:
   `InternalClientLauncher::launch(function<void(wl_display* display)> const&, ...)`

File: include/miral/miral/wayland_extensions.h (hunk 2 above)
Comment (AlanGriffiths) [reply include/miral/miral/wayland_extensions.h]: Simplicity of use and implementation?

File: examples/mir_demo_server/server_example.cpp (hunk 1 above)
Comment (AlanGriffiths) [reply examples/mir_demo_server/server_example.cpp]: I don't see why we need another "Wayland Extensions" object (_Frustra fit per plura quod potest fieri per pauciora)_.

I rejected using a member function...
```c++
        miral::WaylandExtensions{"wl_shell:xdg_wm_base:zxdg_shell_v6:zwlr_layer_shell_v1"}
            .with_extension("xdg_decoration_unstable_v1", [](wl_display*) { return std::shared_ptr<void>{}; })
            .with_extension("some_other_extension", [](wl_display*) { return std::shared_ptr<void>{}; }),
```
... as it seems counter-intuitive to make `with_extension()` a `const` member function.

Comment (AlanGriffiths) [reply examples/mir_demo_server/server_example.cpp]: Also, I had been thinking we simply add these bespoke extensions to the `wayland_extensions` option which gives the shell the opportunity to default them on (and also allows that choice to be overridden). That logic all belongs in `WaylandExtensions`.

That brings up an additional point: the shell ought have a way to validate the actual choice extensions being offered.

**Comment by AlanGriffiths** [comment]
A bigger problem with this API is the lack of an `Executor` to allow Wayland API calls to be made on the correct thread.

**Comment by RAOF** [comment]
It seems like performing tasks on the wayland mainloop would be a common enough task that we'd want to provide an `Executor` to the shell.

**Comment by AlanGriffiths** [comment]
> Looks sensible. No tests, though?

This was just a spike to illustrate an approach. Having established a consensus I'll break the idea out into landable PRs.

**Pull request closed without merging**